src/nanoemoji/bitmap_tables.py

```diff
@@ -49,54 +53,85 @@
 
+        # center within advance
         metrics = BitmapMetrics(
-            x_offset=max(
-                round((_width_in_pixels(config) - config.bitmap_resolution) / 2), 0
+            x_offset=_nudge_into_range(
+                _INT8_RANGE,
+                max(
+                    round(
+                        (
+                            _width_in_pixels(config, image_data)
+                            - config.bitmap_resolution
+                        )
+                        / 2
+                    ),
+                    0,
+                ),
             ),
-            y_offset=round(
-                line_ascent - 0.5 * (line_height - config.bitmap_resolution)
+            y_offset=_nudge_into_range(
+                _INT8_RANGE,
+                round(line_ascent - 0.5 * (line_height - config.bitmap_resolution)),
             ),
             line_height=line_height,
             line_ascent=round(line_ascent),
         )
 
+        # The FontTools errors when values are out of bounds are a bit nasty
+        # so check here for earlier and more helpful termination
+        assert (
+            config.bitmap_resolution in _UINT8_RANGE
+        ), f"bitmap_resolution out of bounds: {config.bitmap_resolution}"
+        assert metrics.y_offset in _INT8_RANGE, f"y_offset out of bounds: {metrics}"
+
         return metrics
 
 
-def _width_in_pixels(config: FontConfig) -> int:
-    return round(
-        config.bitmap_resolution * config.width / (config.ascender - config.descender)
-    )
+def _pixels_to_funits(config: FontConfig, bitmap_pixel_height: int) -> Tuple[int, int]:
+    # the bitmap is vertically scaled to fill the space desc to asc
+    # this gives us a ratio between pixels and upem
+    funits = config.ascender - config.descender
+    return (bitmap_pixel_height, funits)
 
 
-# https://github.com/googlefonts/noto-emoji/blob/9a5261d871451f9b5183c93483cbd68ed916b1e9/third_party/color_emoji/emoji_builder.py#L53
-def _ppem(config: FontConfig, advance: int) -> int:
-    return round(_width_in_pixels(config) * config.upem / advance)
+def _width_in_pixels(config: FontConfig, image_data: PNG) -> int:
+    pixels, funits = _pixels_to_funits(config, image_data.size[1])
 
+    width_funits = image_data.size[0] * funits / pixels
+    width_funits = max(config.width, width_funits)
 
-def _advance(ttfont: ttLib.TTFont, color_glyphs: Sequence[ColorGlyph]) -> int:
-    # let's go ahead and fail miserably if advances are not all the same
-    # proportional bitmaps can wait for a second pass :)
-    advances = {
-        ttfont["hmtx"].metrics[ttfont.getGlyphName(c.glyph_id)][0] for c in color_glyphs
-    }
-    assert len(advances) == 1, "Proportional bitmaps not supported yet"
-    return next(iter(advances))
+    assert width_funits > 0
+    return round(width_funits * pixels / funits)
+
+
+# https://github.com/googlefonts/noto-emoji/blob/9a5261d871451f9b5183c93483cbd68ed916b1e9/third_party/color_emoji/emoji_builder.py#L53
```

Comment:
is the link above still relevant? the code below probably diverged enough to bear little resemblance with it.

Reply:
...no, not it is not. I think perhaps just a link at the top to emoji_builder.py might suffice.